File: client/src/main/kotlin/com/seleneworlds/client/camera/CameraApi.kt
package com.seleneworlds.client.camera

import com.badlogic.gdx.math.Vector3
import com.seleneworlds.common.grid.Coordinate

/**
 * Manage the camera and convert coordinates.
 */
class CameraApi(
    private val cameraManager: CameraManager
) {
    fun getCoordinate(): Coordinate {
        return cameraManager.focusCoordinate
    }

    fun setViewport(x: Int, y: Int, width: Int, height: Int) {
        cameraManager.setViewport(x, y, width, height)
    }

    fun screenToWorld(x: Float, y: Float): Vector3 {
        return cameraManager.unproject(x, y)
    }
}

File: client-web/src/api/CameraApi.ts
import type { Coordinate } from '@/networking/GameProtocol';

export type ClientCameraCoordinateChangedListener = (coordinate: Coordinate) => void;

export interface CameraApi {
  getCoordinate: () => Coordinate;
  setViewport: (x: number, y: number, width: number, height: number) => void;
  screenToWorld: (screenX: number, screenY: number) => { x: number; y: number };
  addCoordinateChangedListener: (listener: ClientCameraCoordinateChangedListener) => () => void;
}
